### PR TITLE
Seperate skill cooldown from trigger cooldown and fix multiple related issues.

### PR DESCRIPTION
**Pull request by Paliak**
Fixes https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/2444 , #4107 , #4210 , #3558, #4959, #4679

Description of the problems being solved:

Separation of Skill cooldown and trigger cooldown in CalcPerform. Correct calculation of triggers for Kitava's Thirst and Craft trigger with improved source selection for the latter. Change Kitava's Thirst to use the in game trigger mod and add parsing for it. Improved handling of hard skill cooldown overrides. Improved handling of a gem being supported by more than one trigger. Adds isTriger property to skills that can trigger other skills. Adds triggeredBy property to skills triggered. Handling of helmet focus trigger (focus trigger now assumes the player recasts focus when the cooldown ends). Correctly account for the fact that some skills have no cooldown in calcMultiSpellRotationImpact. Attack rate calculation for counter attack skills. Fix handling of hard skill cooldown overrides (Discharge). Add support for more unique triggers. Refactoring of trigger code. Improved breakdowns for skills where average trigger rate is affected by accuracy and or crit chance.

The changes are extensive and there may be more edge cases that i was not able to find.

Rundown:

There are flags and properties scattered all over, but the main code starts in CalcActiveSkill.lua 

https://github.com/PathOfBuildingCommunity/PathOfBuilding/blob/e7fccb16b13adeff836494e13079a52bda86989f/src/Modules/CalcActiveSkill.lua#L451-L460

Currently, pob inherits the cooldown from supports when a skill has no cooldown it self. For example, if you link arc to Cast on Crit arc will inherit the cooldown of cast on crit. Later in the code this looks exactly like any other cooldown; Therefore, making it impossible to tell where that cooldown came from (skill itself ? support? cooldown override (discharge)?) This caused issues with cooldown manipulation such as in the case of discharge where the cooldown of the skill itself needs to be overwritten but because both the trigger cooldown and cooldown of the skill get effectively `max()` 'ed there's just no way to do it. This also cases issues when trying to simulate the impact of skill rotation. If multiple skills are linked to a trigger they will be fired off sequentially, but if a skill happens to be on cooldown when the trigger tries to fire it off nothing will happen. @Nostrademous created the `calcMultiSpellRotationImpact` to simulate this so more accurate dps numbers can be provided. The function itself works great, but it was being fed the wrong numbers. You can see the difference that makes in the comment @QuickStick123 has made. It was also missing support for newer trigger mechanics such as lightning conduit adding cast time to cooldown if triggered, and it didn't take cooldown overrides into account. The heart of the code starts where it used to. I didn't touch the mirage archer handling code much, if at all. Simply adapted it to work with the changes I made. Lower starts special handling for the focus trigger. Because it uses the focus mechanic and can trigger all linked skills at the same time getting it to work with the rest of the trigger handling code was difficult, so I decided to make a special if case for it. Going lower, there's the special handling for cast while channeling. Cast while channeling differs from other triggers as it casts the linked skills on a monotonic interval. This means that normal trigger rules don't apply, and thus it has its own section. Just below that is the main trigger handling code. This handles the majority of "normal" triggers. Currently, in pob each trigger has its own section, but because I wanted to add support for more triggers adding a section for every single one quickly turned into copy pasting dozens of lines of code and every small change i made had to be applied and adjusted to every single code block. I decided to create a single block but split it into segments, flexible enough to handle the majority of cases. Stating off, I define some local variables that allow handling specific cases and abstract away some parts to make handling cases such as minion triggers (holy relic) easier. https://github.com/PathOfBuildingCommunity/PathOfBuilding/blob/e7fccb16b13adeff836494e13079a52bda86989f/src/Modules/CalcPerform.lua#L2848-L2862 each skill has different requirements for a triggering skill and can trigger different skills, so it sets different `triggerSkillCond` and `triggerSkillCond` functions that are later used to find and select valid skills. Some triggers also need special handling, such as arcanist brand or global triggers, such as cast on stun or cwdt. If the current skill is not a trigger skill the `skip` flag is set then CalcPerform goes on to do other processing but if a trigger skill is found then first the affected skills need to be found. https://github.com/PathOfBuildingCommunity/PathOfBuilding/blob/e7fccb16b13adeff836494e13079a52bda86989f/src/Modules/CalcPerform.lua#L3079-L3097 This loop uses the functions defined by each trigger to find an appropriate triggering source and to find all other skills that are triggered, so they can be fed into the skill rotation simulation function. Going lower, some validation takes place and the main trigger processing code starts. https://github.com/PathOfBuildingCommunity/PathOfBuilding/blob/e7fccb16b13adeff836494e13079a52bda86989f/src/Modules/CalcPerform.lua#L3110-L3119 First source attack/cast speed modifiers are applied and later the main function that calculates trigger speed is called `calcActualTriggerRate`. This function, along with `getTriggerRateCap` and `calcMultiSpellRotationImpact` calculates the raw trigger rate. It takes into account cooldown recovery and other factors that influence trigger speed. A good chunk of those functions is dedicated to generating a decent looking breakdown for the user as there are so many edge cases.

TODO:
- Implement proper handling for cwdt loops.

Steps taken to verify a working solution:

- Check builds from poe.ninja
- Test build: https://pobb.in/GM-W5q3HkJq7
- Test build2 https://pobb.in/9ADoZbUn-lQK

Live vs pr side by side pictures:

discharge
localidentity
morethanonetrigger
poeninja1
poeninja2

### CODE REVIEW

**Comment by Paliak** [comment]
Found issue. When there's no triggering skill the triggered skill is assumed to be cast by the trigger not self cast. Fixing.

**Comment by Paliak** [comment]
Cast on melee kill and cast on stun working correctly. Should behave closer to CWDT. Fixing.

**Comment by QuickStick123** [comment]
image
image
This skill is supposed to trigger Tawhoa's Chosen which summon's a Mirage Chieftain that attacks once with the given slam that triggered it with no echo/repeat.
image
https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/4732 Will fix this crash by dropping support.

File: src/Data/Skills/sup_str.lua [hunk 1]
```diff
@@ -2794,7 +2796,7 @@ skills["SupportLifetap"] = {
 	support = true,
 	requireSkillTypes = { },
 	addSkillTypes = { SkillType.Duration, },
-	excludeSkillTypes = { SkillType.Blessing, },
```
Comment (QuickStick123) [inline src/Data/Skills/sup_str.lua:2797]: I assume this is just a merge issue

File: src/Modules/CalcOffence.lua [hunk 2]
```diff
@@ -1549,41 +1549,32 @@ function calcs.offence(env, actor, activeSkill)
...
-				activeSkill.skillFlags.HasSeals = true
-			end
 			output.Time = 1 / skillData.triggerRate
 			output.TriggerTime = output.Time
 			output.Speed = skillData.triggerRate
 			skillData.showAverage = false
 		elseif skillData.triggeredByBrand and skillData.triggered then
-			output.Time = 1 / (1 + skillModList:Sum("INC", cfg, "Speed", "BrandActivationFrequency") / 100) / skillModList:More(cfg, "BrandActivationFrequency") * (skillModList:Sum("BASE", cfg, "ArcanistSpellsLinked") or 1)
+			local activationFreqInc = 100 / (100 + skillModList:Sum("INC", cfg, "Speed", "BrandActivationFrequency"))
+			local activationFreqMore = skillModList:More(cfg, "BrandActivationFrequency")
```
Comment (QuickStick123) [inline src/Modules/CalcOffence.lua:1558]: This was divided in the original formula which is correct? (more)

**Comment by QuickStick123** [comment]
There seems to be large differences in simulated impact of spells between master and this I asumme this is intended?
https://pastebin.com/gsBUpSN1
image
image

**Comment by Paliak** [comment]
The simulation incorrectly used the cooldown of the trigger for simulation as the cooldown of the triggered skill (main point of this pr is to separate the values). So basically currently in release the simulator thinks that eye of winter has a cooldown of 0.15s (same as awakened cast on crit it's linked to) in the build you linked, which afaik is incorrect, thus the difference.

File: src/Modules/CalcOffence.lua (hunk 2 above)
Comment (Paliak) [reply src/Modules/CalcOffence.lua:1558]: I've broken the formula down a bit here. division = multiplication by the inverse. It _should_ produce the same values as before. 
obraz

Comment (QuickStick123) [reply src/Modules/CalcOffence.lua:1558]: 44% more activations => smaller time

Comment (Paliak) [reply src/Modules/CalcOffence.lua:1558]: Time is 1/1.44  = 0.69 here. So that checks out ? Or should it be the other way around ? In release it's it's just 1.44 so i guess it should be time = 1.44 without the the inverse.

Comment (Paliak) [reply src/Modules/CalcOffence.lua:1558]: I think i had it the other way around. 
obraz

Comment (QuickStick123) [reply src/Modules/CalcOffence.lua:1558]: Looks good now just got to fix the breakdown.

File: src/Data/Skills/sup_str.lua (hunk 1 above)
Comment (Paliak) [reply src/Data/Skills/sup_str.lua:2797]: Yeah not really sure where that came from. Likely from an old merge.

File: src/Modules/CalcOffence.lua (hunk 2 above)
Comment (Paliak) [reply src/Modules/CalcOffence.lua:1558]: obraz

File: src/Modules/CalcOffence.lua [hunk 3]
```diff
@@ -1549,41 +1549,32 @@ function calcs.offence(env, actor, activeSkill)
...
+			local linkedSpellsCount = skillModList:Sum("BASE", cfg, "ArcanistSpellsLinked") or 1
+			
+			output.Speed = activationFreqMore * linkedSpellsCount * activationFreqInc
+			output.Time = 1 / output.Speed
```
Comment (QuickStick123) [inline src/Modules/CalcOffence.lua:1562]:
```suggestion
			local activationFreqInc = (100 + skillModList:Sum("INC", cfg, "Speed", "BrandActivationFrequency")) / 100
			local activationFreqMore = skillModList:More(cfg, "BrandActivationFrequency")
			local linkedSpellsCount = skillModList:Sum("BASE", cfg, "ArcanistSpellsLinked") or 1
			
			output.Speed = activationFreqInc * activationFreqMore / linkedSpellsCount
			output.Time = 1 / output.Speed
```
This is how it should be according to the original implementation.

Comment (Paliak) [reply src/Modules/CalcOffence.lua]: Applied [d1bfd70](https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/4599/commits/d1bfd705c58102fe6144e123de4162a4b07e0fc8). Modified breakdown to match. Thanks!

**Comment by Lilylicious** [comment]
Currently cast while channeling does not apply the less damage on the gem to the triggered skill.

**Comment by Lilylicious** [comment]
I can't find anything else obviously wrong, so tentative LGTM.

**Comment by Paliak** [comment]
While writing the rundown i did some more testing and i think i've failed to address and issue that exists in current release version of pob. Modifiers to effective source attack speed are applied after skill rotation is taken into consideration. This causes slight differences in effective trigger rate as the simulator assumes that all attacks cause the skills to trigger; Which changes the timing. I will try to fix this as soon as possible.  

If this is to be merged i think it'd be best to merge it into beta for now. That way i could get more feedback from users and it won't break things for the majority of users getting ready for the new league.

**Comment by Nightblade** [comment]
(Not sure what's going on with some of those old spelling errors... you may need to freshen your `dev` snapshot on this branch?)

File: src/Modules/CalcPerform.lua [hunk 4]
```diff
@@ -32,176 +33,486 @@ local function findTriggerSkill(env, skill, source, triggerRate, reqManaCost)
...
+			local br = #skills / m_ceil(skill.cd / data.misc.ServerTickTime) * data.misc.ServerTickTime
+			t_insert(tt1_brs, br)
+			tt1_smallest_br = m_min(tt1_smallest_br, br)
+		end
+	end
+	for _, skill in ipairs(skills) do
+		-- the breaking point, where the trigger time is only constrained by the cooldown time
+		-- before this its its either tt0 or tt1, depending on the skills
+		-- after this the trigger time depends on resonance with the attack speed
+		tt2_br = #skills / (m_ceil(skill.cd / data.misc.ServerTickTime) * data.misc.ServerTickTime) * .8
```
Comment (ProphetLamb) [inline src/Modules/CalcPerform.lua:447]: It is sensible to introduce a library function for the ceil to base function

Comment (Paliak) [reply src/Modules/CalcPerform.lua:447]: Other code uses this style so i stuck with it. Could be worth replacing other instances of this across the code base though,

File: src/Modules/CalcPerform.lua [hunk 5]
```diff
@@ -32,176 +33,486 @@ local function findTriggerSkill(env, skill, source, triggerRate, reqManaCost)
...
-		if time >= next_trigger then
-			while skillRotation[index].next_trig > time do
-				index = (index % #skillRotation) + 1
-				if index == currIndex then
-					wasted = wasted + 1
-					-- Triggers are free from the server tick so cooldown starts at current time
-					next_trigger = time + trigger_increment
-					break
+function calcMultiSpellRotationImpact(env, skills, sourceRate, icdr, triggerCD)
+	local SIMU_RESOLUTION = 32
```
[the remark this answers is not in the document]
Comment (Paliak) [reply src/Modules/CalcPerform.lua:302]: Yeah i was experimenting with it a bit and 2 seemed good enough. This is a remainder from the python code.

**Comment by Paliak** [comment]
The new `calcMultiSpellRotationImpact` still seems to produce wrong values. It's likely still caused by lua being 1 indexed because the order in which gems appear in a group changes the values returned.

**Comment by Paliak** [comment]
Fixed the indexing issues. The lua implementation should now behave like the python one. The 20% error is still concerning though. Would be great if you could take a look @ProphetLamb

**Comment by ProphetLamb** [comment]
> Fixed the indexing issues. The lua implementation should now behave like the python one. The 20% error is still concerning though. Would be great if you could take a look @ProphetLamb 

The reduction in Triggerrate of the fast skill (without cd) is because of missed triggers, where the attack rate doesn't align (exhibits resonance) with the frostbomb skill. In those cases the activation is delayed until the next attack.

In the original simulation this is not considered. Therefore the trigger rate is higher and without the resonant peaks

File: src/Modules/Common.lua [hunk 6]
```diff
@@ -816,3 +817,14 @@ function string:split(sep)
...
+function ceil(x, base)
+	base = base or 1
+	return base * m_ceil(x/base)
+end
+
+-- Ceil function with optional base parameter
+function floor(x, base)
+	base = base or 1
+	return base * m_floor(x/base)
+end
```
Comment (QuickStick123) [inline src/Modules/Common.lua:830]: A function like this already exists it uses number of decimals as the parameter though.
```lua
--- Rounds down a number to the nearest <dec> decimal places
---@param val number
---@param dec number
---@return number
function floor(val, dec)
	if dec then
		local mult = 10 ^ dec
		return m_floor(val * mult + 0.0001) / mult
	else
		return m_floor(val)
	end
end
```

Comment (ProphetLamb) [reply src/Modules/Common.lua:830]: Maybe name it `ceil_b` (ceil to base) to distinguish from the version with decimal precision

Comment (Paliak) [reply src/Modules/Common.lua:818]: I'm not sure how i feel about adding yet another implementation of floor to the code base. I'd need to convert all other instances to maintain consistency too.

Comment (ProphetLamb) [reply src/Modules/Common.lua:818]: Converting the remaining uses would be out of scope imo. But I agree that that 2nd parameter in the new `ceil` function might be mistaken as decimal precision, instead of base. Tho the two are interchangeable since `ceil_b(X, .01) == ceil(X, 2)`. 

Afaik the `ceil(X,base)` function was introduced here, so that should be renamed to `ceil_b(X,base)`.

File: src/Modules/CalcPerform.lua [hunk 7]
```diff
@@ -3141,11 +3141,14 @@ function calcs.perform(env, avoidCache)
 
 			-- Set trigger rate
 			local hits_per_cast = env.player.mainSkill.skillPart == 2 and env.player.mainSkill.activeEffect.srcInstance.skillStageCount or 1
-			output.EffectiveSourceRate = hexCastRate * hits_per_cast
+			output.EffectiveSourceRate = hexCastRate
```
Comment (ha107642) [inline src/Modules/CalcPerform.lua:3144]: I don't think hits_per_cast (or rather, the "stages" input) is included in the calculation at all any more, is it?

Also, I noticed that I used the wrong naming convention for hits_per_cast.

**Comment by QuickStick123** [comment]
Arcanist Brand doesn't seem to work on item granted skills. https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/1065

**Comment by Paliak** [comment]
@QuickStick123 The issue seems to be with all two part skills (Arcanist brand has support part and active part). Putting Arcanist Bramd into one group and any other active spell into another group and socketing both groups into the same item results in the support part of the two part skill not being applied to the active skill in the other group even though it's socketed into the same item. I've fixed it in [1f88b70](https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/4599/commits/1f88b703d116495c27967eccd93effe2152bbd49) but i'm really not sure if that's how it should be done and whether or not it breaks some other interaction.

**Comment by QuickStick123** [comment]
Seems like a pretty small change might be nice to put it in a separate PR so it gets merged quicker?

**Comment by QuickStick123** [comment]
Fixes https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/3815
image
But name is too long.

**Comment by QuickStick123** [comment]
https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/3035 https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/2558 https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/2424 https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/1757
Issues related to triggers I saw if you wanted to look at any of them.

**Comment by QuickStick123** [comment]
> Any Trigger Skills on items that come from a skill gem (such as Rain of Arrows from Lioneye's Paws) now share the same cooldown group, so using the gem version of the skill should put the triggered version on cooldown or vice versa.

Not sure if this affects anything.

**Comment by Paliak** [comment]
I think this pr already does this but i'll have to double check.

**Comment by Paliak** [comment]
The simulation may not be correctly handling cool downs for multiple instances of the same skill in the trigger group. Need to do some more digging into that.

**Comment by QuickStick123** [comment]
There is also https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/5648 as well if you ever wanted to try and get that working. The data is stored now as that PR was merged.

**Comment by Paliak** [comment]
Should be easy to do as i'm already handling that for some skills.

**Comment by atfcdskzezazo** [comment]
These things are easy to handle

**Comment by Paliak** [comment]
Few things to note.

Cool downs across groups are not supported for the same reason CWDT loops are not supported (too many assumptions).

Some triggers require all skills to be in the same group due to the way cross group supports are handled currently. There does not seem to be an easy way of fixing it without throwing graph theory at the problem.

Current simulation of skills in the same group seems to be correct, at least according to my current understanding of trigger skills.

**Comment by QuickStick123** [comment]
Is this in a state where it can be added to the beta branch for testing or not due to WIP stare?

**Comment by Paliak** [comment]
This pr is ready to go into beta at [5a16766](https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/4599/commits/5a16766031995348e9a86520e892917c943445f7) but it has been asked of me to refactor trigger code into a seperate `CalcTriggers.lua` file.

TODO:
- [ ] Move all trigger related code into a separate file
- [ ] Refactor current else if tree into a hash map with a trigger config structure
- [ ] Instead of building on top of existing code, restructure into something that is cleaner and easier to expand.
